metal/resource_metal_device.go

```diff
@@ -649,9 +705,55 @@ func resourceMetalDeviceUpdate(d *schema.ResourceData, meta interface{}) error {
 		}
 
 	}
+
+	if d.HasChange("operating_system") || d.HasChange("user_data") || d.HasChange("custom_data") {
+		reinstallOptions, err := getReinstallOptions(d)
+
+		if err != nil {
+			return friendlyError(err)
+		}
+
+		if _, err := client.Devices.Reinstall(d.Id(), &reinstallOptions); err != nil {
+			return friendlyError(err)
+		}
+	}
+
 	return resourceMetalDeviceRead(d, meta)
 }
 
+func getReinstallOptions(d *schema.ResourceData) (packngo.DeviceReinstallFields, error) {
+	reinstall_list, ok := d.Get("reinstall").([]interface{})
+
+	if !ok {
+		return packngo.DeviceReinstallFields{}, fmt.Errorf("expected reinstall configuration and none available")
+	}
+
+	if len(reinstall_list) == 0 {
+		return packngo.DeviceReinstallFields{}, fmt.Errorf("expected reinstall configuration and none available")
+	}
+
+	reinstall_config, ok := reinstall_list[0].(map[string]interface{})
+
+	if !ok {
+		return packngo.DeviceReinstallFields{}, fmt.Errorf("expected reinstall configuration and none available")
+	}
+
+	preserveData, ok := reinstall_config["preserve_data"].(bool)
```

Comment:
The field is called `persist_data` (not `preserve_data`).

The `if !ok { ... = false }` looks to be redundant, we would only need this condition to optionally set a `DeviceReinstallFields` field where the absence of the field (nil or omitted) would have a different meaning than `false`.